### ISSUE TITLE
Let Dumpling set GC safepoint ttl with new pd interface

### ISSUE THREAD

**Issue by lichunzhu** [pingcap/dumpling#95]
When Dumpling dumps large data it needs to set TiDB snapshot at a ts. When the dumping time is too long tikv may gc this TiDB snapshot ts and Dumpling will fail to dump, exits with an error.

To avoid this problem, users need to set `tikv_gc_life_time` longer, which is explained in the end of https://pingcap.com/docs-cn/stable/export-or-backup-using-dumpling. However, it's not convenient. The users need to run
```sql
update mysql.tidb set VARIABLE_VALUE = '720h' where VARIABLE_NAME = 'tikv_gc_life_time';
```
before dumping and run
```sql
update mysql.tidb set VARIABLE_VALUE = '10m' where VARIABLE_NAME = 'tikv_gc_life_time';
```
after dumping.

But, Dumpling shouldn't change this variable because dumping process may fail in the half.

In v4.0.0, PD has a new interface [UpdateServiceGCSafePoint](https://github.com/pingcap/pd/blob/master/client/client.go#L82) which allows service to set TiKV GC safe point with a ttl, which is very suitable for dumpling's scene. What's more, TiCDC has already used this interface [in their service](https://github.com/pingcap/ticdc/blob/514a16d6b76b08b3a5b7ddb8cc853e1065852246/cdc/owner.go#L418).

So, in summary, I think dumpling can also use this feature to make sure dumped ts is not "GCed".

**Comment by kennytm** [comment]
(I'm setting this P3 because TiDB backup should primarily use BR rather than Dumpling.)

**Comment by lichunzhu** [comment]
> (I'm setting this P3 because TiDB backup should primarily use BR rather than Dumpling.)

Okay

**Comment by kennytm** [comment]
Concern: Dumpling and the TiDB cluster may be in different intranets. When TiDB tells us that PD is on 10.0.0.3:2379, Dumpling may be connecting to a totally irrelevant PD.

![Dumpling connecting to TiDB through a reverse proxy](https://user-images.githubusercontent.com/103023/84111668-02dc6c00-aa5a-11ea-8465-3fe1b963c924.png)

**Comment by lichunzhu** [comment]
> Concern: Dumpling and the TiDB cluster may be in different intranets. When TiDB tells us that PD is on 10.0.0.3:2379, Dumpling may be connecting to a totally irrelevant PD.
> 
> ![Dumpling connecting to TiDB through a reverse proxy](https://user-images.githubusercontent.com/103023/84111668-02dc6c00-aa5a-11ea-8465-3fe1b963c924.png)

Yes... It sames that lightning will also meet this problem?

**Comment by kennytm** [comment]
Lightning (Importer and Local backend) won't have this problem since Lightning requires direct connection with PD and TiKV.

Lightning (TiDB/Loader backend) doesn't connect to PD at all.

**Comment by lichunzhu** [comment]
> Lightning (Importer and Local backend) won't have this problem since Lightning requires direct connection with PD and TiKV.
> 
> Lightning (TiDB/Loader backend) doesn't connect to PD at all.

Lightning will automatically get pd address through statusPort/settings when it's not given. Is it the same as here?

**Comment by kennytm** [comment]
> Lightning will automatically get pd address through statusPort/settings when it's not given. Is it the same as here?

Lightning won't automatically get PD addresses when backend = TiDB.

**Comment by lichunzhu** [comment]
But the given PD addresses may also be local, am I right?
IMO, we can leave a warning for users in this situation. In this case users must change GC ts by themselves

**Comment by kennytm** [comment]
In Lightning with backend = TiDB, if you pass the --pd-urls it will just be ignored since Lightning in tidb backend doesn't have any action involving PD.

The problem in #97 is Dumpling will *unconditionally* `UpdateServiceGCSafePoint` using only PD address fetched from TiDB.

**Comment by lichunzhu** [comment]
I see... I think the best solution is current handle in #97 -- ignore this error. Any better advise?

If TiDB gives a local PD address but dumpling connects to another cluster through this PD address. Hard to imagine this situation..

**Comment by kennytm** [comment]
well imagine copying data from one cluster to another which involves changes that BR cannot handle (e.g. modifying new-collation)

**Comment by lichunzhu** [comment]
![Dumpling connecting to TiDB through a reverse proxy](https://user-images.githubusercontent.com/103023/84111668-02dc6c00-aa5a-11ea-8465-3fe1b963c924.png)

To solve this problem that getted pd and connected TiDB are from different network cluster depicted like the above pitcute, we must add another check to make sure getted pd and connected 
TiDB are in one cluster.

One Solution:
Use **the ddl_id of TiDB** (denoted by UUID to assure the uniquity) which can be both queryed from pd and TiDB:

### From PD:
can be found from etcd:
```bash
/tidb/server/info/461dafe8-ef1b-44be-8d47-459c1402396f
{"version":"5.7.25-TiDB-v4.0.0","git_hash":"689a6b6439ae7835947fcaccf329a3fc303986cb","ddl_id":"461dafe8-ef1b-44be-8d47-459c1402396f","ip":"127.0.0.1","listening_port":4002,"status_port":10082,"lease":"45s","binlog_status":"Off","start_timestamp":1591692033}
/tidb/server/info/75bf2409-7ae6-4e98-9c76-afe92b019db1
{"version":"5.7.25-TiDB-v4.0.0","git_hash":"689a6b6439ae7835947fcaccf329a3fc303986cb","ddl_id":"75bf2409-7ae6-4e98-9c76-afe92b019db1","ip":"127.0.0.1","listening_port":4000,"status_port":10080,"lease":"45s","binlog_status":"Off","start_timestamp":1591692032}
/tidb/server/info/f11e79e9-a4dc-434d-8738-b58de37ee40d
{"version":"5.7.25-TiDB-v4.0.0","git_hash":"689a6b6439ae7835947fcaccf329a3fc303986cb","ddl_id":"f11e79e9-a4dc-434d-8738-b58de37ee40d","ip":"127.0.0.1","listening_port":4001,"status_port":10081,"lease":"45s","binlog_status":"Off","start_timestamp":1591692032}
```
### From TiDB:
can be found in table `INFORMATION_SCHEMA`.`TIDB_SERVERS_INFO`:
```sql
+--------------------------------------+-----------+------+-------------+-------+--------------------+------------------------------------------+---------------+
| DDL_ID                               | IP        | PORT | STATUS_PORT | LEASE | VERSION            | GIT_HASH                                 | BINLOG_STATUS |
+--------------------------------------+-----------+------+-------------+-------+--------------------+------------------------------------------+---------------+
| f11e79e9-a4dc-434d-8738-b58de37ee40d | 127.0.0.1 | 4001 |       10081 | 45s   | 5.7.25-TiDB-v4.0.0 | 689a6b6439ae7835947fcaccf329a3fc303986cb | Off           |
| 461dafe8-ef1b-44be-8d47-459c1402396f | 127.0.0.1 | 4002 |       10082 | 45s   | 5.7.25-TiDB-v4.0.0 | 689a6b6439ae7835947fcaccf329a3fc303986cb | Off           |
| 75bf2409-7ae6-4e98-9c76-afe92b019db1 | 127.0.0.1 | 4000 |       10080 | 45s   | 5.7.25-TiDB-v4.0.0 | 689a6b6439ae7835947fcaccf329a3fc303986cb | Off           |
+--------------------------------------+-----------+------+-------------+-------+--------------------+------------------------------------------+---------------+
```

**Comment by GregoryIan** [comment]
how about `cluster_id`?

**Comment by lichunzhu** [comment]
> how about `cluster_id`?

We can query cluster_id from PD. But according to TiDB developers TiDB doesn't support query clusterID through TiDB SQL now.
